### PR TITLE
Support launching flutter_tools from a package URI

### PR DESCRIPTION
**Pull request by jason-simmons**

### CODE REVIEW

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart [hunk 1]
```diff
@@ -183,6 +183,12 @@ class FlutterCommandRunner extends CommandRunner<void> {
     try {
       if (platform.script.scheme == 'data')
         return '../..'; // we're running as a test
+
+      if (platform.script.scheme == 'package') {
+        String packageConfigPath = Uri.parse(platform.packageConfig).toFilePath();
```
Comment (aam) [inline packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: Perhaps give an error if packageConfig is null (when no `--packages` flag is specified)?

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart [hunk 2]
```diff
@@ -183,6 +183,12 @@ class FlutterCommandRunner extends CommandRunner<void> {
...
+        return fs.path.dirname(fs.path.dirname(fs.path.dirname(packageConfigPath)));
```
Comment (aam) [inline packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: Perhaps add a comment regarding directory structure we expect? I'm not sure what this navigation three levels up mean? Given 'package:flutter_gallery/main.dart' we go three levels up from .packages file, is that right?

**Review by aam — APPROVED** [review]

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart (hunk 1 above)
Comment (jason-simmons) [reply packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: If there is no `packageConfig`, then Dart would not have been able to launch `flutter_tools` if a `package:` URI was specified on the Dart command line

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart (hunk 2 above)
Comment (jason-simmons) [reply packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: The paths in `_defaultFlutterRoot` are based on the directory structure of `packages/flutter_tools` in the Flutter framework repository.  The packageConfig for `flutter_tools` lives at `packages/flutter_tools/.packages`, so the Flutter root is three levels up from there.

**Pull request merged**